### PR TITLE
Subscribe certain labels only (white list)

### PR DESCRIPTION
**Pull request by cou929**
refs #384

- Implemented a feature to register white list of labels to subscribe with reference to https://github.com/integrations/slack/issues/384#issuecomment-376593013 and https://github.com/integrations/slack/issues/384#issuecomment-377368011
- Next I want to implement the feature to register labels as black list if this pull request has no problem
    - I will create another pull request for this

I am a bit worried about whether the command syntax is appropriate and implementation is properly. So I want feedback.

## Command syntax

```
# Notify events which has `todo` label
/github subscribe owner/repo label:todo

# Notify events that have at least one `todo` or `help wanted` label
# You can specify labels which has spaces when enclosed in quotes
/github subscribe owner/repo label:todo label:"help wanted"

# Unsubscribe
/github unsubscribe owner/repo label:todo label:"help wanted"
```

- You can specify label as white list
- If multiple labels are specified, it considered as `OR` condition

## Implementation notes

- Command syntax has become complicated, so I modified it to parse with regular expressions
- `label` are saved in `Subscriptions.settings` as json format such like `{ label: [...] }`
- In `lib/activity/router.js`, compares `Subscriptions.settings.label` and label of the webhook payload and filter it

### CODE REVIEW

**Comment by cou929** [comment]
I've added some changes to improve testing coverage. And force pushed it for readability of commit logs.
Could you provide feedback? Thanks.

**Comment by cou929** [comment]
resolved conflict ca8c70e

**Comment by cou929** [comment]
resolved conflict f6bd6df

File: lib/activity/router.js [hunk 1]
```diff
@@ -76,6 +76,38 @@ module.exports = function route(callback) {
...
+        const shouldFilterByLabel = (() => {
+          if (issue === undefined) {
+            return false;
+          }
+
+          if (!Array.isArray(issue.labels)) {
+            return false;
+          }
+
+          if (subscription.settings.label === undefined) {
```
Comment (padamczewski) [inline lib/activity/router.js]:
```suggestion
          if (subscription.settings && subscription.settings.label === undefined) {
```

Is there guarantee that `subscription.settings` is available at this point? If it's not then you might get `Uncaught TypeError: Cannot read property of undefined`

**Comment by scf4** [comment]
Any update on this?

**Comment by ArturKlajnerok** [comment]
Hi @wilhelmklopp & @gimenete, this is a highly requested feature: https://github.com/integrations/slack/issues/384 and can improve the workflow for many teams. 
Could you consider to prioritize the review for this PR? Let me know if I can help to push this forward.

**Comment by sojeri** [comment]
Hi @scepticulous & @wilhelmklopp -- I am also very invested in seeing this PR or a similar one go in which addresses #384. Is there anything I or those above can do to help this PR get reviewed or otherwise be considered ready to check in?

**Comment by webdif** [comment]
> Next I want to implement the feature to register labels as black list if this pull request has no problem. I will create another pull request for this

It would be so nice to be able to blacklist label with a `-`, for example `/github subscribe owner/repo -label:standby` 😍

**Comment by cakism** [comment]
Really looking forward to this feature, hoping to see it live soon! <3

**Comment by Joncy** [comment]
Up: this is extremely useful for large repos with many contributors.

**Comment by leog** [comment]
Would love to have this as well! A review from someone with write access is required: @gimenete @wilhelmklopp are you up to it? 💪

**Comment by webdif** [comment]
How can we help? I would looove to see this feature merged 😄

**Comment by nmurthy** [comment]
+1 would love to help push this one through

**Comment by scepticulous** [comment]
Hey @cou929 👋 ,
my changes and changes to master overlapped and therefore my changes are no longer associated with this PR. Not sure how to fix this easily, without conflicts, so I decided to create a new PR that is based on the awesome work you did here. Thanks again 🙇 

Closes in favor of #911

**Comment by ahmad380360** [comment]
شكرا لجميع

**Pull request closed without merging**